### PR TITLE
[AdminBundle] OAuthUserCreator Should query on username and email

### PR DESCRIPTION
**Pull request by Numkil**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets | #1144 

When someone signs in for the first time using google sign in, his user should be searched based on
email or username.

### CODE REVIEW

**Comment by Numkil** [comment]
I just realised there could be still a problem with this code in the following use case. 
There are 2 users in the database. One with username == [the email we search for] and one with email == [the email we search for]. That way it would be impossible to decide which user should be logged in...

**Comment by jockri** [comment]
Perhaps we should have logic that it first searches for matching email. If not found, also search username. What do you think?

**Comment by Numkil** [comment]
I gave find by email a higher priority as find by username. That way if there are 2 accounts present with that email address the one with the email address in the email field will be chosen.

**Comment by mlebkowski** [comment]
How about giving the developer a choice? I.e. extract some kind `UserFinderInterface`, create a default service implementing it with the above logic, but with the possibility for the developer to replace it with his own implementation (replacing a service in the container), hmm?

**Comment by Numkil** [comment]
@mlebkowski following your advice I pulled everything apart and set everything behind interfaces. This should allow other people to override either only the code for finding a user based on the googleId and email address or the complete login procedure. 

@jockri I think this should be quite ready now

File: src/Kunstmaan/AdminBundle/Helper/Security/OAuth/OAuthUserCreatorInterface.php [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+     * Checks if there already exists an account for the given googleId or email. If yes updates
+     * the access levels accordingly and returns that user. If no creates a new user with the
+     * configured access levels.
+     *
+     * Returns Null if email is not in configured domains
+     *
+     * @param string email
+     * @param string googleId
+     *
+     * @return AbstractUser Implementation
```
Comment (mlebkowski) [inline src/Kunstmaan/AdminBundle/Helper/Security/OAuth/OAuthUserCreatorInterface.php]: `@return mixed AbstractUser Implementation` — this way IDE won’t expect to return `AbstractUser` instance. Maybe it should be a `UserInterface` instance from FOS?

File: src/Kunstmaan/AdminBundle/Helper/Security/OAuth/OAuthUserFinder.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    /** @var string */
+    private $userClass;
+
+    /**
+     * OAuthUserCreator constructor.
+     * @param EntityManager $em
+     * @param $userClass
+     */
+    public function __construct(EntityManager $em, $userClass)
```
Comment (mlebkowski) [inline src/Kunstmaan/AdminBundle/Helper/Security/OAuth/OAuthUserFinder.php]: `EntityManagerInterface`

File: src/Kunstmaan/AdminBundle/Helper/Security/OAuth/OAuthUserFinderInterface.php [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+{
+
+    /**
+     * Tries to find a user in database based on email and googleId fields.
+     * Returns null when nothing has been found.
+     *
+     * @param string email
+     * @param string googleId
+     *
+     * @return AbstractUser Implementation
```
Comment (mlebkowski) [inline src/Kunstmaan/AdminBundle/Helper/Security/OAuth/OAuthUserFinderInterface.php]: same here, `mixed`

**Pull request merged**